hazelcast/src/main/java/com/hazelcast/sql/impl/client/SqlClientService.java

```diff
@@ -88,27 +89,52 @@ public SqlResult execute(@Nonnull SqlStatement statement) {
                 statement.getTimeoutMillis(),
                 statement.getCursorBufferSize(),
                 statement.getSchema(),
-                SqlClientUtils.expectedResultTypeToByte(statement.getExpectedResultType())
+                SqlClientUtils.expectedResultTypeToByte(statement.getExpectedResultType()),
+                id
             );
 
-            ClientMessage responseMessage = invoke(requestMessage, connection);
+            SqlClientResult res = new SqlClientResult(
+                this,
+                connection,
+                id,
+                statement.getCursorBufferSize()
+            );
+
+            ClientInvocationFuture future = invokeAsync(requestMessage, connection);
+
+            future.whenComplete((message, error) -> handleExecuteResponse(connection, res, message, error));
 
-            SqlExecute2Codec.ResponseParameters response = SqlExecute2Codec.decodeResponse(responseMessage);
+            return res;
+        } catch (Exception e) {
+            throw rethrow(e, connection);
+        }
+    }
+
+    private void handleExecuteResponse(
+        ClientConnection connection,
+        SqlClientResult res,
+        ClientMessage message,
+        Throwable error
+    ) {
+        if (error != null) {
+            res.onExecuteError(rethrow(error, connection));
+
+            return;
+        }
+
+        try {
```

Comment:
Is try-catch block necessary? As far as I understand the only place that could throw an exception is `handleResponseError`. And we can set that exception on `SqlClientResult res` without throwing. 
Similarly, we can avoid throwing and catching exceptions in `handleFetchResponse`.
The same could be applied to `fetch` even though we can not remove try-catch.

Reply:
Fixed.